src/net/connector.rs

```diff
@@ -33,11 +41,20 @@ impl UdpConnector {
 
 impl NetworkConnector for UdpConnector {
     type Stream = UdpSender;
-    
+
     fn connect(&self, host: &str, port: u16, _: &str) -> error::Result<<Self as NetworkConnector>::Stream> {
         let udp_sock = try!(self.0.try_clone());
-        let sock_addr = try!(net::addr_from_trait((host, port)));
-        
+        let sock_addr = match try!(self.local_addr()) {
+            SocketAddr::V4(_) => SocketAddr::V4(SocketAddrV4::new(FromStr::from_str(host).unwrap(), port)),
+            SocketAddr::V6(n) => {
+                let mut addr: SocketAddrV6 = FromStr::from_str(format!("{}:{}", host, port).as_str())
```

Comment:
Tricky problem. I run into that panic and is why I changed the code there.

The thing is _hyper_ is "incorrectly" splitting the address from the port (by incorrectly, I mean, doing it old school Ipv4 days). 

The ipv6 address "[x:y:z]:8123" is being sent as `connect("[x:y:z]", 8123, _)` **incorrect** instead of `connect("x:y:z", 8123)`.
 
The "[]" are not part of an address, they are there to separate address from port. 
Since the "[]" where already present in host, I had to parse format!("{}:{}", host, port) into a `SocketAddr` that can understand the "[]" because attempting to parse "[x:y:z]" as an IpAddr fails.

I guess we could check if the host is enclosed in `"[" + inner + "]"` and based on that do:
`format!("{}:{}", host, port).as_str())` or `format!("[{}]:{}", host, port).as_str())`